ratis-logservice/src/main/java/org/apache/ratis/logservice/server/LogStateMachine.java

```diff
@@ -452,29 +503,166 @@ public void close() {
-       if (this.state != state) {
-          return new IOException("Wrong state: " + this.state);
-        }
+  private Throwable verifyState(State... states) {
+    for (State state : states) {
+      if (this.state == state) {
         return null;
-   }
+      }
+    }
+    return new IOException("Wrong state: " + this.state);
+  }
 
+  private CompletableFuture<Message> updateArchiveLogInfo(
+      LogServiceRequestProto logServiceRequestProto) {
+    LogServiceProtos.ArchiveLogRequestProto archiveLog = logServiceRequestProto.getArchiveLog();
+    this.archiveLogName = LogServiceProtoUtil.toLogName(archiveLog.getLogName());
+    this.archiveLocation = archiveLog.getLocation();
+    this.lastArchivedIndex = archiveLog.getLastArchivedRaftIndex();
+    Throwable t = verifyState(State.ARCHIVING);
+    return CompletableFuture.completedFuture(
+        Message.valueOf(LogServiceProtoUtil.toArchiveLogReplyProto(t).toByteString()));
+  }
+
+  private CompletableFuture<Message> processArchiveLog(
+      LogServiceRequestProto logServiceRequestProto) {
+    LogServiceProtos.ArchiveLogRequestProto archiveLog = logServiceRequestProto.getArchiveLog();
+    LogName logName = LogServiceProtoUtil.toLogName(archiveLog.getLogName());
+    String loc = null;
+    this.isArchival = !archiveLog.getIsExport();
+    if(isArchival) {
+      loc = archiveLocation;
+    }else{
+      loc = archiveLog.getLocation();
+    }
+    if (loc == null) {
+      throw new IllegalArgumentException(isArchival ?
+          "Location for archive is not configured" :
+          "Location for export provided is null");
+    }
+    final String location = loc;
+    String archiveLocationForLog = LogServiceUtils.getArchiveLocationForLog(location, logName);
+    long recordId = archiveLog.getLastArchivedRaftIndex();
+    try {
+      Throwable t = verifyState(State.CLOSED);
+      if (t == null) {
+        Callable<Boolean> callable = () -> {
+            sendChangeStateRequest(State.ARCHIVING);
+            updateArchivingInfo(recordId, logName, location, isArchival);
+            ArchiveLogWriter writer = new ArchiveHdfsLogWriter();
+            writer.init(archiveLocationForLog,logName);
+            LogServiceRaftLogReader reader = new LogServiceRaftLogReader(log);
+            reader.seek(0);
+            long records = 0;
+            while (reader.hasNext()) {
+              writer.write(ByteBuffer.wrap(reader.next()));
+              if (records >= DEFAULT_ARCHIVE_THRESHOLD_PER_FILE || isNoMoreLeader) {
+                commit(writer, logName, location);
+                if (isNoMoreLeader) {
+                  break;
+                }
+                records = 0;
+              }
+              records++;
+            }
+            writer.close();
+            if (!isNoMoreLeader) {
+              sendChangeStateRequest(State.ARCHIVED);
+            } else {
+              sendArchiveLogrequestToNewLeader(writer.getLastWrittenRecordId(), logName, location);
```

Comment:
In the case that the active leader of a LogServer quorum bounces around a lot, we're going to be "chaining" query() RPCs, aren't we? While we wait on computation of archive in "this" instance, we might trigger another archive in other instance, which triggers another archive in a third instance, etc. I'd worry about this cascading through the cluster in a leader election storm.

Would having the metadata quorum drive this archiving logic be a way to try to prevent that? Maybe there are more problems in trying to do that which I'm not thinking about?

Reply:
That's a good point on scalability. Sounds good to try to prevent the bouncing leader problem :)